### ISSUE TITLE
[Epic] Bring up Tegra CUDA Toolkit

### ISSUE THREAD

**Issue by traversaro** [conda-forge/cuda-feedstock#57]
### Comment:

Hello @conda-forge/cuda,

I work in robotics R&D, and in my field it is widespread the use of Jetson Boards. Until now, for compatibility with non-Nvidia hardware we were forced to use Jetpack 5, and so CUDA 11, so we  could not use the latest CUDA-enabled packages from conda-forge (as we do in most of the rest of our installations).

A bit naively, some days ago we just tried to install `pytorch` on the Jetson Board, and we experienced an error (https://github.com/conda-forge/pytorch-cpu-feedstock/issues/303) that was related to the missing `8.7` compute capabilities in the conda-forge's `pytorch` build. At that point, we locally built a new `pytorch` package with `8.7` capabilities enabled. After trying that, we had similar problems on `libcublas` and `libcudnn`. By forcing the use of the `apt` version of `libcublas` and `libcudnn` (that we verified matched the version of the conda-installed version), `pytorch` started to working fine.

After that, I started to investigated how to enabled `cuda` packages that could work fine on Jetson Orin, and I discovered the difference between `sbsa` and `tegra`, and all the related machinery in conda. Sorry for the noise in the PRs I opened in this investigation. : )

As suggested in https://github.com/conda-forge/libcublas-feedstock/pull/32#issuecomment-2537155195, I open this issue in here to discuss the possibility of having official CUDA packages with 8.7 compute capability enabled, and from what I understand this means having packages that could be installed that are compatible with `arm-variant=*=*tegra*`.

I have a few doubts on this, that I think can be used to kickstart the discussion:
* D1: The CUDA 12 `tegra` build of `cudnn` and `cublas` contains code for `sm_62` and `sm_72` architectures. However, from what I understand (from https://developer.nvidia.com/cuda-gpus) those architectures are only available in Jetson Xavier and Jetson TX2 boards, that if I am not wrong can't install CUDA 12 as they do not support the Jetpack 6. This is not really a question for cuda package itself, I was just wondering if it could make sense for downstream packages in conda-forge to enable or not the  `sm_62` and `sm_72` architectures.
* D2: At the moment, there is no mechanism to automatically understand which `arm-variant` should be installed on a given system (see https://github.com/conda-forge/arm-variant-feedstock/issues/3). For me, that is not a blocker, as even if `conda create -n pytorchenv pytorch-gpu` does not work out of the box on a Jetson (as the `sbsa` variant would get installed out of the box), it is a satisfactory solution having to run `conda create -n pytorchenv pytorch-gpu arm-variant=*=*tegra*` rather then not having working packages at all. However, what is most important for now is avoiding that `sbsa` users get a `tegra` package installed accidentally. I just tried to install `arm-variant` in conda and pixi and indeed the `sbsa` version gets installed, but I am not sure if that is robust. Should we enforce someway that `sbsa` packages have higher priority w.r.t. to `tegra` ones?
* D3: Does the `tegra` builds have different minimum glibc requirements? I had some test failures in cudnn recipe (`"/lib64/libm.so.6: version 'GLIBC_2.29' not found (required by [..]/libcudnn_engines_runtime_compiled.so)"`), not sure if this is something that needs to be accounted for.

Related issues and draft PRs:
* https://github.com/conda-forge/pytorch-cpu-feedstock/issues/303
* https://github.com/conda-forge/cusparselt-feedstock/issues/48#issuecomment-2535670765
* https://github.com/conda-forge/cuda-cudart-feedstock/pull/38
* https://github.com/conda-forge/cuda-nvrtc-feedstock/pull/19
* https://github.com/conda-forge/cuda-nvtx-feedstock/pull/26
* https://github.com/conda-forge/cudnn-feedstock/pull/100
* https://github.com/conda-forge/libcublas-feedstock/pull/32
* https://github.com/conda-forge/arm-variant-feedstock/issues/3

**Comment by carterbox** [comment]
[Caveats for Tegra systems](https://docs.nvidia.com/cuda/cuda-for-tegra-appnote/#cuda-features-not-supported-on-tegra)

**Comment by carterbox** [comment]
[This table shows which devices are supported by each Jetpack version.](https://developer.nvidia.com/embedded/jetpack-archive)

[This table shows which CTK versions are supported by each Jetpack version.](https://docs.nvidia.com/cuda/cuda-for-tegra-appnote/index.html#use-the-right-upgrade-package)

[This table shows which devices are which compute capabilities.](https://developer.nvidia.com/cuda-gpus)

**Comment by carterbox** [comment]
When attempting to prototype the compiling workflow for a third-party package to target tegra, we ran into some issues. So we have determined that we probably need a second metapackage to indicate that we want to "build for" tegra.

Currently we have the `arm-variant` meta package which lets users select which redists are installed at runtime. This implies that the meaning of `arm-variant` is “packages which depend on “arm-variant=tegra” run on tegra devices.” This seems to work well enough. We have sbsa variants of packages have a high priority and so users can select to switch to tegra variants by installing `arm-variant`=tegra.

However, what happens when we try to compile on conda-forge? We lack a mechanism to distinguish between target architecture and build architecture. For example:
```yaml
build:
  # This package means the target is arm, but which variant?
  - cuda-nvcc_linux-aarch64
  # How do you specify that the target is tegra?
  # We cannot do this:
  # - arm-variant * tegra
  # because that would mean the compiler runs on tegra not
  # that it targets tegra
host:
  # means that packages here run on {{ arm_variant_type }}
  - arm-variant * {{ arm_variant_type }}
run:
  # means that packages here run on {{ arm_variant_type }}
  - arm-variant * {{ arm_variant_type }}
```
All compiling of tegra packages on conda-forge must be cross-compiled because there are no tegra runners. Which means we must have a method to distinguish between target and build arm variant. We cannot use the `arm-variant` package alone because this package means a run dependency on an architecture. If we put this package in the build environment, then since compilers are also binaries, this would mean that the compilers themselves run on tegra, but this is not the case.

Therefore, we need a second meta-package something like: `cuda-nvcc-arm-variant-target` which a compiler which targets arm can use to detect whether it should compile for sbsa or tegra.

This would result in something like the following

build:
  - cuda-nvcc_linux-aarch64
  - arm-variant-target * {{ arm_variant_type }}
host:
  - arm-variant * {{ arm_variant_type }}
run:
  - arm-variant * {{ arm_variant_type }}



xref: https://docs.google.com/document/d/1Kks6DBSaED6Uz3qh3epIkoyUG7iKWXrjtsf9_p6rzaQ/edit?usp=sharing

**Comment by carterbox** [comment]
- [ ] https://github.com/conda-forge/cuda-cccl-feedstock/pull/21
- [ ] https://github.com/conda-forge/cuda-nvcc-impl-feedstock/pull/35
- [ ] https://github.com/conda-forge/cuda-nvcc-feedstock/pull/57
- [ ] https://github.com/conda-forge/arm-variant-feedstock/pull/9
- [ ] https://github.com/conda-forge/cuda-cudart-feedstock/pull/38

**Comment by traversaro** [comment]
Thanks a lot for all the effort @carterbox !

**Comment by traversaro** [comment]
An example of user interest in this feature: https://github.com/prefix-dev/pixi/issues/3122 .

**Comment by traversaro** [comment]
Hello @carterbox, thanks a lot for all the work on the Nvidia side on this. There is anything on the community side we can do to help on this?

**Comment by carterbox** [comment]
@traversaro, not yet. We will need help testing third-party feedstocks once we have published tegra conda-packages for the CTK. I'll ping you when that is complete.

**Comment by traversaro** [comment]
> [@traversaro](https://github.com/traversaro), not yet. We will need help testing third-party feedstocks once we have published tegra conda-packages for the CTK. I'll ping you when that is complete.

Thanks a lot for the update!

**Comment by carterbox** [comment]
For reference, CTK 13.0 has done some unification work for tegra/sbsa, but it applies to new architectures only.

https://docs.nvidia.com/cuda/archive/13.0.0/cuda-toolkit-release-notes/index.html#id2

**Comment by billysuh7** [comment]
I'd like to get approval for https://github.com/conda-forge/cuda-nvrtc-feedstock/pull/28 and https://github.com/conda-forge/libnvjitlink-feedstock/pull/31 in order to set up PR for `libcublas`, `libcusolver` and `libcusparse`. Thanks!

**Comment by carterbox** [comment]
Note to self: We probably only support Jetpack 6.x devices because our minimum glibc requirement will be 2.34, and Jetpack 5.x uses Ubuntu 20.04 which does not meet this glibc requirement.

**Comment by billysuh7** [comment]
I'd appreciate if someone could review https://github.com/conda-forge/libcusparse-feedstock/pull/29 , as this is a dependency for `libcusolver` which I have yet to raise PR for. Thanks!

**Comment by billysuh7** [comment]
Ugh, sorry, `libcublas` PR also needs to be merged before `libcusolver` PR is setup. @carterbox could you approve https://github.com/conda-forge/libcublas-feedstock/pull/45 as well ?

**Comment by traversaro** [comment]
> [@traversaro](https://github.com/traversaro), it looks like it's now possible to install the cuda-toolkit for tegra. e.g.
> 
> ```
> conda create -n test --platform linux-aarch64 arm-variant=*=tegra cuda-toolkit cuda-version=12.9
> ```
> 
> Do you have some packages you can build and test?

Great! I think two packages that we use on orin architecture that we can easily test (and that do not depend on cudnn) are https://github.com/conda-forge/librealsense-feedstock and https://github.com/conda-forge/llama.cpp-feedstock, let me open a PR for those.

**Comment by traversaro** [comment]
llama.cpp PR: https://github.com/conda-forge/llama.cpp-feedstock/pull/75

**Comment by traversaro** [comment]
Ok, I reached a good point on https://github.com/conda-forge/librealsense-feedstock/pull/79 and https://github.com/conda-forge/llama.cpp-feedstock/pull/75 . The llama.cpp required a few non-tegra fixes, so I spinned those in https://github.com/conda-forge/llama.cpp-feedstock/pull/76, but the tegra build of llama.cpp is working fine. To test it, I uploaded it on a temporary channel, and a simple benchmark of llama can be run with:

~~~bash
git clone https://github.com/traversaro/pixi-working-examples
cd pixi-working-examples/llama-cpp-jetson-orin
pixi info
pixi run benchmark
~~~

I do not have access to a Jetson Orin with JetPack 6 at the moment, but I asked some colleagues to perform the test tomorrow. 

I noticed that there is no cudnn package for `arm-variant=*=tegra`, there is any specific blocking issue on that?

**Comment by carterbox** [comment]
Note to self: Jetpack 5.x devices (sm_72 Xavier) can only support CTK<=12.2. To support these devices we would need to also republish the CTK for 12.2 and then use cuda minor version compatibility when building. We probably won't do this because conda-forge only builds for CUDA>=12.9 now.

**Comment by traversaro** [comment]
> Ok, I reached a good point on [conda-forge/librealsense-feedstock#79](https://github.com/conda-forge/librealsense-feedstock/pull/79) and [conda-forge/llama.cpp-feedstock#75](https://github.com/conda-forge/llama.cpp-feedstock/pull/75) . The llama.cpp required a few non-tegra fixes, so I spinned those in [conda-forge/llama.cpp-feedstock#76](https://github.com/conda-forge/llama.cpp-feedstock/pull/76), but the tegra build of llama.cpp is working fine. To test it, I uploaded it on a temporary channel, and a simple benchmark of llama can be run with:
> 
> git clone https://github.com/traversaro/pixi-working-examples
> cd pixi-working-examples/llama-cpp-jetson-orin
> pixi info
> pixi run benchmark
> 
> I do not have access to a Jetson Orin with JetPack 6 at the moment, but I asked some colleagues to perform the test tomorrow.
> 
> I noticed that there is no cudnn package for `arm-variant=*=tegra`, there is any specific blocking issue on that?

Thanks to @giotherobot we tested this on a Jetson Orin, and we confirmed that is working fine:

~~~
System
------------
       Pixi version: 0.58.0
           Platform: linux-aarch64
   Virtual packages: __unix=0=0
                   : __linux=5.15.148=0
                   : __glibc=2.35=0
                   : __cuda=12.6=0
                   : __archspec=1=cortex_a72
          Cache dir: /home/ergocub/.cache/rattler/cache
       Auth storage: /home/ergocub/.rattler/credentials.json
   Config locations: No config files found

ergocub@ergocub-head:~/pixi-working-examples/llama-cpp-jetson-orin$ pixi run benchmark
✨ Pixi task (download_model in default): curl -L -o gemma-3-1b-it-Q4_K_M.gguf https://huggingface.co/ggml-org/gemma-3-1b-it-GGUF/resolve/main/gemma-3-1b-it-Q4_K_M.gguf                                    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1335  100  1335    0     0   5826      0 --:--:-- --:--:-- --:--:--  5855
100  768M  100  768M    0     0  39.1M      0  0:00:19  0:00:19 --:--:-- 40.2M
                                                                                                     ✨ Pixi task (benchmark in default): llama-bench -m ./gemma-3-1b-it-Q4_K_M.gguf                      ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: Orin, compute capability 8.7, VMM: yes
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| gemma3 1B Q4_K - Medium        | 762.49 MiB |   999.89 M | CUDA       |  99 |           pp512 |        935.19 ± 5.68 |
| gemma3 1B Q4_K - Medium        | 762.49 MiB |   999.89 M | CUDA       |  99 |           tg128 |         17.83 ± 0.02 |

build: 3cae8ea (158)
~~~

**Comment by traversaro** [comment]
Personally I never fully understood how `track_features` work (or are supposed to work), especially across the different `mamba` and `rattler` solvers (or also `classic`, but perhaps we can forget about it). I always had more luck with build numbers, and it seems that most recipes also handle higher priority of CUDA variant with higher build numbers.

**Comment by carterbox** [comment]
@traversaro, The YAML migrator is now live, and [our documentation](https://github.com/conda-forge/cuda-feedstock/blob/main/recipe/doc/recipe_guide.md#building-for-arm-tegra-devices) on how to build for Tegra are updated. You or any other downstream package maintainer can now safely begin building packages for Tegra.

Some in-flight PRs that you may be interested in include:

- https://github.com/conda-forge/cusparselt-feedstock/pull/74
- https://github.com/conda-forge/cudnn-feedstock/pull/129
- https://github.com/conda-forge/libmagma-feedstock/pull/34
- https://github.com/conda-forge/nccl-feedstock/pull/151

**Comment by jakirkham** [comment]
Thank you Daniel for all of your hard work in bringing this up! 🙏

Also thanks to everyone trying these packages out and providing us feedback. Will be interested to see how broader adoption goes
